jfuzzylite/src/test/java/com/fuzzylite/rule/RuleWeightTest.java

```diff
@@ -63,14 +47,18 @@ public void testAddRuleWithWeightInRuleText() throws Exception {
     }
 
     @Test
-    public void testRuleChangeWeight() throws Exception {
+    public void testRuleWeightChange() throws Exception {
```

Comment:
The reason why I tested these was that if we do learning, which is basically adding / removing rules and changing weight of existing rules, it would be more convenient to manipulate the rules in code instead of do some string manipulation on rule strings. So in that sense I would be more keen on fixing this than removing it from the API.

Reply:
Thank you for pointing this out. I totally agree with you. How about we delete the constructor with weight, and we write a bit of javadoc warning the reset of the weight on loading the rule?

Reply:
What if we change the Rule(str, w) to basically call this(str + " with " + w)? Maybe with check that rule does not already contain 'with' and throw an exception in that case?

I did not quite get you on the javadoc warning part though?

Reply:
I like your idea, but I am not keen one bit on having a constructor throwing an exception, especially thinking of the behaviour in C++. I am still thinking on what we could do, but still leaning towards the deletion of constructor. On the javadoc, I meant just writing up that the weight of the rule will be set automatically upon reloading to whatever the value is on the rule text, or something like that. 
Could you try your use case deleting the constructor and let me know how it feels?

Reply:
I'm fine with the deletion if we get setWeight  (or if it's already)  to work as expected.

I think my current approach on learning would be creating an Engine from existing fll. Doing some manipulation with the rules (add, remove, change weight) and the exporting it back to fll where the actual program that does the fuzzything can pick it up.

Reply:
I will take a closer look at this during the week. Just a thought, if we have a method in Rule called ‘update()’, we could replace in the rule text “with x” to the current weight value w or append “with w” to the rule if it does not have a weight. As such, we have the base rule text, a changing weight, and we update the base rule text when the weight is appropriate. Still with this, the constructor with weight is to be deleted. Thoughts?

Reply:
Would it be enough to add the weight only when doing an export? I find it confusing that weight is some what semi parsed from string to weight but not utilized in all the places. What I mean is that the engine logic should use the rule.getWeight() for all it's needs and only export would add `" with " + rule.getWeight()` (forgive me as do not know the codebase so well that what I'm saying here might already be true)

Reply:
Hey, thanks again for your comments, and please excuse me for the delay. I like what you suggest, so given this case:
`rule: if A then Z with W=1.0`, we load the rule, we change the weight for learning, we export it as you suggest, and we reload the rule without modifying its weight. However, this also means that we need to modify the Rule::parse method so we exclude the "with X" from all the rules (if present). And also means that, when exporting, we need to add the `with {Op::str(w)}` only if w!=1.0.

Still, there is something I am not fully convinced. What do we do with getText()? Should it contain "with w"? I am more inclined to adding a new method `Rule::text() -> return _text + " with " + w`?

Reply:
And there is another bit too:
- When we export using Op::str(w), we will be rounding/truncating to whatever decimals the library has set. This will affect the results when we use a rule using a truncated weight. Alternatively, we could write the whole "w", but the result will surely not be nice.

Reply:
Just to keep our options open, I will argue in favour of the option I proposed earlier.
- We keep the rule the way it is, and we add a method called `update()` (or something similar), which appends ("with w") or replaces the entire value "w" (not using `Op::str()`). 

I like about this that (1) the text always represents the initial rule and will be parsed and exported exactly how it was initially written; (2) there is no complication on what we are rounding and how nicely we present weights; (3) it is simple and backward compatible; and (4) provides a way to do learning with the weights knowing what the initial state was. Think of the `update` method as a way to persist any experimental weight you play with.

Reply:
AFAIK `Rule#weight` is only used in `Rule#activateWith`. So if I call setWeight between calls to `Engine#process` the weight given will be utilized by the engine but if the engine is exported it will probably contain the wrong weight. FisExporter seems to be the only exporter that calls ´Rule#getWeight` and others seem to rely on `Rule#getText()`.